### DISCUSSION TITLE
How I can use "appname://some?option=val" in edge to launch my app?

### DISCUSSION THREAD

**Discussion by BTMuli** [tauri-apps/tauri#7844] [Q&A] [answered]
I want my application to achieve such functionality: it can be invoked through a third party/Url and accept some parameters.

At first, I was studying how to make the packaged app.exe accept parameters when calling. Later I learned that Cutsom Protocol can implement URL calling applications.

I've seen discussions involving Custom Protocol, and compared the use of the localhost plugin, but these confused me even more: how do I modify the configuration to allow the custom protocol, how do I write main.rs to accept parameters to start/revoke the application, and How to accept parameters in frontend (Node).

I also tried the deep-link plugin, but it didn't work.

**Comment by FabianLars** [comment]
> I also tried the deep-link plugin, but it didn't work.

That's unfortunate because that's exactly what you're asking for! Any specific issues you hit that you can share so we can figure out what went wrong?

**Reply by BTMuli** [reply]
Here's the Cargo.toml
```toml
[dependencies]
tauri = { version = "1.4", features = ["api-all"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
tauri-plugin-deep-link="0.1.2"
```

The main.rs
```rust
use tauri::Manager;

fn main() {
    tauri_plugin_deep_link::prepare("TeyvatGuide");
    tauri::Builder::default()
        .plugin(tauri_plugin_sql::Builder::default().build())
        .setup(|_app| {
          let handle = _app.handle();
          tauri_plugin_deep_link::register(
                "my-scheme",
                move |request| {
                  dbg!(&request);
                  handle.emit_all("test_deep_link", request).unwrap();
                },
              ).unwrap();
            let _window = _app.get_window("TeyvatGuide").unwrap();
            #[cfg(debug_assertions)] // only include this code on debug builds
                _window.open_devtools(); // open the devtools on startup
            Ok(())
        })
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

The front App.vue
```typescript
async function getDeepLink(): Promise<void> {
  await event.listen("test_deep_link", e=>{
    console.log(e.payload)
  });
}
```


```text
[✔] Environment
    - OS: Windows 10.0.22621 X64
    ✔ WebView2: 116.0.1938.81
    ✔ MSVC: Visual Studio Community 2022
    ✔ rustc: 1.70.0 (90c541806 2023-05-31)
    ✔ Cargo: 1.70.0 (ec8a8a0ca 2023-04-25)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 18.16.0
    - pnpm: 8.7.5
    - npm: 9.8.1

[-] Packages
    - tauri [RUST]: 1.4.1
    - tauri-build [RUST]: 1.4.0
    - wry [RUST]: 0.24.4
    - tao [RUST]: 0.16.2
    - @tauri-apps/api [NPM]: 1.4.0
    - @tauri-apps/cli [NPM]: 1.4.0

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist
    - devPath: http://localhost:3000/
    - framework: Vue.js
    - bundler: Vite
```

**Reply by BTMuli** [reply]
And when I put "TeyvatGuide://test" in edge，there is nothing.

**Reply by FabianLars** [reply, marked as answer]
Okay, first of all you're still registering my-scheme as the link scheme, so in case that's not just a copy paste error on github you have to change that.
Secondly, you may have to use all lowercase on the browser side (you can use any casing to register it tho) -> What i mean is that even tho you register it as `TeyvatGuide` the browser will redirect that to `teyvatguide` which may cause problems so may as well enter `teyvatguide` in the browser directly.
Lastly, if you let the link start your app, the event you send to the frontend will be lost since the listener won't be registered at that point.

( All the above was from testing your code on my end. After applying my suggestions it works for me now. )

**Reply by BTMuli** [reply]
> Lastly, if you let the link start your app, the event you send to the frontend will be lost since the listener won't be registered at that point.

Thanks,I try to register it after setup with invoke:
```rust

#[tauri::command]
async fn register_deep_link(app_handle: tauri::AppHandle) {
    tauri_plugin_deep_link::register(
        "teyvat_guide",
        move |request| {
            dbg!(&request);
            app_handle.emit_all("test_deep_link", request).unwrap();
        },
    )
    .unwrap();
}

fn main() {
    tauri_plugin_deep_link::prepare("teyvat_guide");
    tauri::Builder::default()
        .plugin(tauri_plugin_sql::Builder::default().build())
        .setup(|_app| {
            let _window = _app.get_window("TeyvatGuide").unwrap();
            #[cfg(debug_assertions)] // only include this code on debug builds
                _window.open_devtools(); // open the devtools on startup
            Ok(())
        })
        .invoke_handler(tauri::generate_handler![register_deep_link])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

and in App.vue
```typescript
onMounted(async () => {
  await tauri.invoke("register_deep_link");
  await getDeepLink();
});

async function getDeepLink(): Promise<void> {
  await event.listen("test_deep_link", (e) => {
    console.log(e.payload);
  });
}
```
But when I input "teyvat_guide://something", there still nothing console in devtool,is it where wrong ?

> ( All the above was from testing your code on my end. After applying my suggestions it works for me now. )

I'm interested how it works,what's the code the what uri input in your edge.

Rust is difficult for me since I just use tauri to bundle a msi,so something will be a hard for me while it's basic for ruster.ORZ

**Reply by FabianLars** [reply]
`_` (underscores) are not allowed in the scheme in a browser. But in this case it will also be automatically replaced. So i copy pasted your code as-is (with `teyvat_guide`) but had to use `teyvat-guide://something` in edge.

Do you see any errors in the console where you run `tauri dev`?

**Reply by BTMuli** [reply]
Ok,I replace "teyvatguide" with "teyvat_guide" and magically,it works!
![image](https://github.com/tauri-apps/tauri/assets/72692909/ea590753-90eb-4488-9fe4-da679a35b811)
